prov/util/src/rocr_mem_monitor.c

```diff
@@ -0,0 +1,378 @@
+
+	if (ofi_iov_left(&entry->iov, iov))
+		return -1;
+	else if (ofi_iov_right(&entry->iov, iov))
+		return 1;
+
+	/* If this fails, the ROCR memory monitor failed to have a single ROCR
+	 * memory monitor entry per user allocated ROCR buffer.
+	 */
+	assert(ofi_iov_within(iov, &entry->iov));
+
+	return 0;
+}
+
+static struct rocr_mm_entry *rocr_mm_entry_find(const void *addr, size_t len)
+{
+	struct ofi_rbnode *node;
+	struct iovec iov = {
+		.iov_base = (void *) addr,
+		.iov_len = len,
+	};
+
+	node = ofi_rbmap_find(&rocr_mm.entries, (void *) &iov);
+	if (node)
+		return node->data;
+
+	return NULL;
+}
+
+/* Pointer to ROCR memory monitor entry can never be returned as user data. This
+ * could lead to use-after-free. Instead, address and length is always returned.
+ * Unsubscribe will attempt to lookup the corresponding ROCR memory monitor
+ * entry and will free the entry if found.
+ */
+static void rocr_mm_dealloc_cb(void *addr, void *user_data)
+{
+	size_t len = (size_t) user_data;
+
+	pthread_mutex_lock(&mm_lock);
+	ofi_monitor_notify(rocr_monitor, addr, len);
+	ofi_monitor_unsubscribe(rocr_monitor, addr, len, NULL);
+	pthread_mutex_unlock(&mm_lock);
+}
+
+static void rocr_mm_entry_free(struct rocr_mm_entry *entry)
+{
+	const void *addr = entry->iov.iov_base;
+	size_t len = entry->iov.iov_len;
+	hsa_status_t hsa_ret;
+
+	FI_DBG(&core_prov, FI_LOG_MR,
+	       "ROCR buffer address %p length %lu monitor entry freed\n", addr,
+	       len);
+
+	/* Two return codes are expected. HSA_STATUS_SUCCESS is returned if the
+	 * deallocation callback was not triggered and the entry is freed.
+	 * HSA_STATUS_ERROR_INVALID_ARGUMENT is returned if the deallocation
+	 * callback was triggered and the entry is freed. Any other return code
+	 * puts the monitor in an unknown state and is fatal.
+	 */
+	hsa_ret = ofi_hsa_amd_deregister_deallocation_callback(entry->iov.iov_base,
+							       rocr_mm_dealloc_cb);
+	assert(hsa_ret == HSA_STATUS_SUCCESS ||
+	       hsa_ret == HSA_STATUS_ERROR_INVALID_ARGUMENT);
+
+	ofi_rbmap_delete(&rocr_mm.entries, entry->node);
+	dlist_remove(&entry->entry);
+	free(entry);
+}
+
+static int rocr_mm_entry_alloc(const void *addr, size_t len,
+			       struct rocr_mm_entry **entry)
+{
+	hsa_amd_pointer_info_t hsa_info;
+	hsa_status_t hsa_ret;
+	int ret;
+
+	*entry = malloc(sizeof(**entry));
+	if (!*entry) {
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	/* Determine full ROCR memory region size. */
+	hsa_ret = ofi_hsa_amd_pointer_info((void *) addr, &hsa_info, NULL, NULL,
+					   NULL);
+	if (hsa_ret != HSA_STATUS_SUCCESS) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Failed to perform hsa_amd_pointer_info: %s\n",
+			ofi_hsa_status_to_string(hsa_ret));
+
+		ret = -FI_EIO;
+		goto err_free_entry;
+	}
+
+	/* Requested memory length must be within the allocated length. If not,
+	 * user may be trying to monitor unallocated memory.
+	 */
+	if ((uintptr_t) addr + len >
+	    (uintptr_t) hsa_info.agentBaseAddress + hsa_info.sizeInBytes) {
```

Comment:
Can the address range specified by the application span more than one device region?  The code as written suggests that cannot be the case.  Is that true?

Reply:
This should be removed. I was able to allocate a MR spanning to separate but virtually contiguous memory allocations.